### ISSUE TITLE
etcd Version: 3.4.29 panic: runtime error: invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by TurtleSnail** [etcd-io/etcd#17265]
### Bug report criteria

- [ ] This bug report is not security related, security issues should be disclosed privately via security@etcd.io.
- [ ] This is not a support request or question, support requests or questions should be raised in the etcd [discussion forums](https://github.com/etcd-io/etcd/discussions).
- [ ] You have read the etcd [bug reporting guidelines](https://github.com/etcd-io/etcd/blob/main/Documentation/contributor-guide/reporting_bugs.md).
- [ ] Existing open issues along with etcd [frequently asked questions](https://etcd.io/docs/latest/faq) have been checked and this is not a duplicate.

### What happened?

When running as non-root user etcd segments
```
/usr/local/bin/etcd   --auto-compaction-retention 2   --auto-compaction-mode revision   --cert-file=/etc/etcd/certs/etcd.pem   --key-file=/etc/etcd/certs/etcd-key.pem   --client-cert-auth   --trusted-ca-file=/etc/etcd/certs/etcd-ca.pem   --listen-client-urls=https://127.0.0.1:2379   --advertise-client-urls=https://127.0.0.1:2379
WARNING: Package "github.com/golang/protobuf/protoc-gen-go/generator" is deprecated.
	A future release of golang/protobuf will delete this package,
	which has long been excluded from the compatibility promise.

[WARNING] Deprecated '--logger=capnslog' flag is set; use '--logger=zap' flag instead
2024-01-17 13:05:01.658422 W | embed: Running http and grpc server on single port. This is not recommended for production.
2024-01-17 13:05:01.658488 I | etcdmain: etcd Version: 3.4.29
2024-01-17 13:05:01.658503 I | etcdmain: Git SHA: 3ee15038f
2024-01-17 13:05:01.658508 I | etcdmain: Go Version: go1.20.12
2024-01-17 13:05:01.658548 I | etcdmain: Go OS/Arch: linux/amd64
2024-01-17 13:05:01.658558 I | etcdmain: setting maximum number of CPUs to 2, total number of available CPUs is 2
2024-01-17 13:05:01.658573 W | etcdmain: no data-dir provided, using default data-dir ./default.etcd
2024-01-17 13:05:01.658707 N | etcdmain: the server is already initialized as member before, starting as etcd member...
[WARNING] Deprecated '--logger=capnslog' flag is set; use '--logger=zap' flag instead
2024-01-17 13:05:01.658745 W | embed: Running http and grpc server on single port. This is not recommended for production.
2024-01-17 13:05:01.660575 I | embed: name = default
2024-01-17 13:05:01.660610 I | embed: data dir = default.etcd
2024-01-17 13:05:01.660628 I | embed: member dir = default.etcd/member
2024-01-17 13:05:01.660639 I | embed: heartbeat = 100ms
2024-01-17 13:05:01.660648 I | embed: election = 1000ms
2024-01-17 13:05:01.660658 I | embed: snapshot count = 100000
2024-01-17 13:05:01.660698 I | embed: advertise client URLs = https://127.0.0.1:2379
2024-01-17 13:05:01.660713 I | embed: initial advertise peer URLs = http://localhost:2380
2024-01-17 13:05:01.660725 I | embed: initial cluster = 
2024-01-17 13:05:01.663309 I | etcdserver: restarting member 8e9e05c52164694d in cluster cdf818194e3a8c32 at commit index 12
raft2024/01/17 13:05:01 INFO: 8e9e05c52164694d switched to configuration voters=()
raft2024/01/17 13:05:01 INFO: 8e9e05c52164694d became follower at term 6
raft2024/01/17 13:05:01 INFO: newRaft 8e9e05c52164694d [peers: [], term: 6, commit: 12, applied: 0, lastindex: 12, lastterm: 6]
2024-01-17 13:05:01.664870 W | auth: simple token is not cryptographically signed
2024-01-17 13:05:01.669029 I | etcdserver: starting server... [version: 3.4.29, cluster version: to_be_decided]
2024-01-17 13:05:01.669131 I | pkg/fileutil: started to purge file, dir: default.etcd/member/snap, suffix: snap.db, max: 5, interval: 30s
2024-01-17 13:05:01.669218 I | pkg/fileutil: started to purge file, dir: default.etcd/member/snap, suffix: snap, max: 5, interval: 30s
2024-01-17 13:05:01.669320 I | pkg/fileutil: started to purge file, dir: default.etcd/member/wal, suffix: wal, max: 5, interval: 30s
raft2024/01/17 13:05:01 INFO: 8e9e05c52164694d switched to configuration voters=(10276657743932975437)
2024-01-17 13:05:01.670051 I | etcdserver/membership: added member 8e9e05c52164694d [http://localhost:2380] to cluster cdf818194e3a8c32
2024-01-17 13:05:01.670340 N | etcdserver/membership: set the initial cluster version to 3.4
2024-01-17 13:05:01.670457 I | etcdserver/api: enabled capabilities for version 3.4
2024-01-17 13:05:01.671337 I | embed: ClientTLS: cert = /etc/etcd/certs/etcd.pem, key = /etc/etcd/certs/etcd-key.pem, trusted-ca = /etc/etcd/certs/etcd-ca.pem, client-cert-auth = true, crl-file = 
2024-01-17 13:05:01.671539 I | embed: listening for peers on 127.0.0.1:2380
raft2024/01/17 13:05:03 INFO: 8e9e05c52164694d is starting a new election at term 6
raft2024/01/17 13:05:03 INFO: 8e9e05c52164694d became candidate at term 7
raft2024/01/17 13:05:03 INFO: 8e9e05c52164694d received MsgVoteResp from 8e9e05c52164694d at term 7
raft2024/01/17 13:05:03 INFO: 8e9e05c52164694d became leader at term 7
raft2024/01/17 13:05:03 INFO: raft.node: 8e9e05c52164694d elected leader 8e9e05c52164694d at term 7
2024-01-17 13:05:03.276396 I | etcdserver: published {Name:default ClientURLs:[https://127.0.0.1:2379]} to cluster cdf818194e3a8c32
2024-01-17 13:05:03.276573 I | embed: ready to serve client requests
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x75845d]

goroutine 173 [running]:
go.uber.org/zap.(*Logger).check(0x0, 0x2, {0x101d390, 0x16})
	/root/go/pkg/mod/go.uber.org/zap@v1.10.0/logger.go:264 +0x7d
go.uber.org/zap.(*Logger).Error(0xc0001e2400?, {0x101d390?, 0x0?}, {0xc0000a5b40, 0x1, 0x1})
	/root/go/pkg/mod/go.uber.org/zap@v1.10.0/logger.go:202 +0x3e
go.etcd.io/etcd/embed.(*serveCtx).serve(0xc0001e2400, 0xc0003aac00, 0xc0003aa768, {0x118ede0, 0xc000032880}, 0xc00002ef20, 0xc000026c60?, 0x0, {0xc00002c5c0, 0x2, ...})
	/tmp/etcd-release-3.4.29/etcd/release/etcd/embed/serve.go:121 +0x507
go.etcd.io/etcd/embed.(*Etcd).serveClients.func1(0xc0000e09c0?)
	/tmp/etcd-release-3.4.29/etcd/release/etcd/embed/etcd.go:806 +0x105
created by go.etcd.io/etcd/embed.(*Etcd).serveClients
	/tmp/etcd-release-3.4.29/etcd/release/etcd/embed/etcd.go:805 +0xa35
```

### What did you expect to happen?

not segement, or give some error message

### How can we reproduce it (as minimally and precisely as possible)?

download the binary, generate certs, run the command.

### Anything else we need to know?

ubuntu 22.04

### Etcd version (please run commands below)

<details>

```console

$ etcd --version
WARNING: Package "github.com/golang/protobuf/protoc-gen-go/generator" is deprecated.
	A future release of golang/protobuf will delete this package,
	which has long been excluded from the compatibility promise.

etcd Version: 3.4.29
Git SHA: 3ee15038f
Go Version: go1.20.12
Go OS/Arch: linux/amd64

$ etcdctl version
$ etcdctl version
etcdctl version: 3.4.29
API version: 3.4
```

</details>

### Etcd configuration (command line flags or environment variables)

<details>

# paste your configuration here

</details>

### Etcd debug information (please run commands below, feel free to obfuscate the IP address or FQDN in the output)

<details>

```console
$ etcdctl member list -w table
# paste output here

$ etcdctl --endpoints=<member list> endpoint status -w table
# paste output here
```

</details>

**Comment by ahrtr** [comment]
Thanks for raising the issue and sorry for the inconvenience. 

Fixed in https://github.com/etcd-io/etcd/pull/17270

**Comment by ahrtr** [comment]
But you also need to investigate why `registerGateway` failed. Please raise a separate PR if you think it's an etcd bug and provide the log,

https://github.com/etcd-io/etcd/blob/071cb4319002789de9699f8cc2dd52defe0f7a5c/embed/serve.go#L119-L120

**Comment by ahrtr** [comment]
Resolved in https://github.com/etcd-io/etcd/pull/17270. The fix will be included in 3.4.30
